pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/rancher/wrangler/pkg/generic/fake"
+	"github.com/stretchr/testify/require"
+	admissionv1 "k8s.io/api/admission/v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+)
+
+const (
+	nsName = "test"
+	user   = "test-user"
+)
+
+func TestAdmit(t *testing.T) {
+	workspace := v3.FleetWorkspace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nsName,
+		},
+	}
+	raw, _ := json.Marshal(workspace)
+	req := &admission.Request{
+		AdmissionRequest: admissionv1.AdmissionRequest{
+			Object: runtime.RawExtension{Object: &workspace, Raw: raw},
+			UserInfo: authenticationv1.UserInfo{
+				Username: user,
+			},
+		},
+	}
+
+	tests := map[string]struct {
+		m                  func(t *testing.T) Mutator
+		req                *admission.Request
+		expectAllowed      bool
+		expectResultStatus *metav1.Status
+		expectedErr        error
+	}{
+		"reject because namespace already exists": {
+			m:             nsExistMutator,
+			req:           req,
+			expectAllowed: false,
+			expectResultStatus: &metav1.Status{
+				Status:  "Failure",
+				Message: "namespace 'test' already exists",
+				Reason:  metav1.StatusReasonBadRequest,
+				Code:    http.StatusBadRequest,
+			},
+		},
+		"new namespace without the 'app.kubernetes.io/managed-by: rancher' label": {
+			m:                  newNsMutator,
+			req:                req,
+			expectAllowed:      true,
+			expectResultStatus: nil,
+		},
+		"existing namespace with the 'app.kubernetes.io/managed-by: rancher' label": {
+			m:                  newNsWithLabelAndValidPermissionsMutator,
+			req:                req,
+			expectAllowed:      true,
+			expectResultStatus: nil,
+		},
+		"reject existing namespace with the 'app.kubernetes.io/managed-by: rancher' label with invalid permissions": {
+			m:             newNsWithLabelAndInvalidPermissionsMutator,
+			req:           req,
+			expectAllowed: false,
+			expectResultStatus: &metav1.Status{
+				Message: `user "test-user" (groups=[]) is attempting to grant RBAC permissions not currently held:
+{APIGroups:["management.cattle.io"], Resources:["fleetworkspaces"], ResourceNames:["test"], Verbs:["*"]}`,
+				Status: "Failure",
+				Reason: metav1.StatusReasonForbidden,
+				Code:   http.StatusForbidden,
+			},
+		},
+		"reject because namespace can't be fetched": {
+			m:           newNsErrorMutator,
+			req:         req,
+			expectedErr: errors.NewServerTimeout(schema.GroupResource{}, "", 2),
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			m := test.m(t)
+			admit, err := m.Admit(test.req)
+			if err != nil {
+				require.Equalf(t, test.expectedErr, err, "Error doesn't match, wanted %v got %v", test.expectedErr, err)
+				return
+			}
```

Comment:
Making the check on the returned error, instead of the expected one, may expose us to unnoticed failures if no error is returned although one is expected.
Conversely, running equality checks only if the expected error is not `nil` means that if no error is expected but one is returned, it may fly under our radar.

TL;DR:
```suggestion
			if test.expectedErr == nil {
				require.NoError(t, err)
		        } else {
				require.Equalf(t, err, "Error doesn't match, wanted %v got %v", test.expectedErr, err)
			}
```

Reply:
changed here https://github.com/rancher/webhook/pull/260/commits/fc284bd621528d5128f03b4573f4c60b05afc6dc